File: Kernel Simulator/Arguments/ArgumentParse.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Module ArgumentParse

    'Variables
    Public argcommands As String
    Public argcmds() As String

    ''' <summary>
    ''' Parses specified arguments
    ''' </summary>
    Public Sub ParseArguments()

        'Check for the arguments written by the user
        Try
            For i As Integer = 0 To EnteredArguments.Count - 1
                Dim indexArg As Integer = EnteredArguments(i).IndexOf(" ")
                If indexArg = -1 Then
                    indexArg = EnteredArguments(i).Count
                    EnteredArguments(i) = EnteredArguments(i).Substring(0, indexArg)
                End If
                If AvailableArgs.Contains(EnteredArguments(i).Substring(0, indexArg)) Then
                    If EnteredArguments(i) = "quiet" Then

                        DefConsoleOut = Console.Out
                        Console.SetOut(IO.StreamWriter.Null)

                    ElseIf EnteredArguments(i).Contains("cmdinject") Then

                        'Command Injector argument
                        If EnteredArguments(i) = "cmdinject" Then
                            Write(DoTranslation("Available commands: {0}") + vbNewLine +
                              DoTranslation("Write command: "), False, ColTypes.Input, String.Join(", ", Commands.Keys))
                            argcmds = Console.ReadLine().Split({" : "}, StringSplitOptions.RemoveEmptyEntries)
                            argcommands = String.Join(", ", argcmds)
                            If argcommands <> "q" Then
                                CommandFlag = True
                            Else
                                Write(DoTranslation("Command injection has been cancelled."), True, ColTypes.Neutral)
                            End If
                        Else
                            argcmds = EnteredArguments(i).Substring(10).Split({" : "}, StringSplitOptions.RemoveEmptyEntries)
                            argcommands = String.Join(", ", argcmds)
                            CommandFlag = True
                        End If

                    ElseIf EnteredArguments(i) = "debug" Then

                        DebugMode = True

                    ElseIf EnteredArguments(i) = "maintenance" Then

                        maintenance = True

                    ElseIf EnteredArguments(i) = "safe" Then

                        SafeMode = True

                    ElseIf EnteredArguments(i) = "testInteractive" Then

                        InitTShell()
                        If Test_ShutdownFlag Then Environment.Exit(0)

                    End If
                Else
                    Write(DoTranslation("bargs: The requested argument {0} is not found."), True, ColTypes.Error, EnteredArguments(i).Substring(0, indexArg))
                End If
            Next
        Catch ex As Exception
            KernelError("U", True, 5, DoTranslation("bargs: Unrecoverable error in argument: ") + ex.Message, ex)
        End Try

    End Sub

End Module

File: public/Kernel Simulator/Arguments/ArgumentBase/ArgumentParse.cs

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using KS.Arguments.CommandLineArguments;
using KS.Kernel;
using KS.Kernel.Debugging;
using KS.Kernel.Exceptions;
using KS.Languages;
using KS.Misc.Writers.ConsoleWriters;
using KS.Shell.ShellBase.Commands;
using System;
using System.Collections.Generic;

namespace KS.Arguments.ArgumentBase
{
    /// <summary>
    /// Argument parser class
    /// </summary>
    public static class ArgumentParse
    {

        /// <summary>
        /// Available command line arguments
        /// </summary>
        public readonly static Dictionary<string, ArgumentInfo> AvailableCMDLineArgs = new()
        {
            { "quiet", new ArgumentInfo("quiet", "Starts the kernel quietly", new CommandArgumentInfo(), new QuietArgument()) },
            { "maintenance", new ArgumentInfo("maintenance", "Like safe mode, but also disables multi-user and some customization", new CommandArgumentInfo(), new MaintenanceArgument()) },
            { "safe", new ArgumentInfo("safe", "Starts the kernel in safe mode, disabling all mods", new CommandArgumentInfo(), new SafeArgument()) },
            { "testInteractive", new ArgumentInfo("testInteractive", "Opens a test shell", new CommandArgumentInfo(), new CommandLine_TestInteractiveArgument()) },
            { "debug", new ArgumentInfo("debug", "Enables debug mode", new CommandArgumentInfo(), new CommandLine_DebugArgument()) },
            { "terminaldebug", new ArgumentInfo("terminaldebug", "Enables terminal debug mode", new CommandArgumentInfo(), new CommandLine_TerminalDebugArgument()) },
            { "reset", new ArgumentInfo("reset", "Resets the kernel to the factory settings", new CommandArgumentInfo(), new CommandLine_ResetArgument()) },
            { "bypasssizedetection", new ArgumentInfo("bypasssizedetection", "Bypasses the console size detection", new CommandArgumentInfo(), new CommandLine_BypassSizeDetectionArgument()) },
            { "help", new ArgumentInfo("help", "Help page", new CommandArgumentInfo(), new CommandLine_HelpArgument()) }
        };

        /// <summary>
        /// Parses specified arguments
        /// </summary>
        /// <param name="ArgumentsInput">Input Arguments</param>
        public static void ParseArguments(List<string> ArgumentsInput)
        {
            // Check for the arguments written by the user
            try
            {
                var Arguments = AvailableCMDLineArgs;

                // Parse them now
                for (int i = 0; i <= ArgumentsInput.Count - 1; i++)
                {
                    string Argument = ArgumentsInput[i];
                    if (Arguments.ContainsKey(Argument))
                    {
                        // Variables
                        var ArgumentInfo = new ProvidedArgumentArgumentsInfo(Argument);
                        var Args = ArgumentInfo.ArgumentsList;
                        var Switches = ArgumentInfo.SwitchesList;
                        string strArgs = ArgumentInfo.ArgumentsText;
                        bool RequiredArgumentsProvided = ArgumentInfo.RequiredArgumentsProvided;

                        // If there are enough arguments provided, execute. Otherwise, fail with not enough arguments.
                        if (Arguments[Argument].ArgArgumentInfo.ArgumentsRequired & RequiredArgumentsProvided | !Arguments[Argument].ArgArgumentInfo.ArgumentsRequired)
                        {
                            var ArgumentBase = Arguments[Argument].ArgumentBase;
                            ArgumentBase.Execute(strArgs, Args, Switches);
                        }
                        else
                        {
                            DebugWriter.WriteDebug(DebugLevel.W, "User hasn't provided enough arguments for {0}", Argument);
                            TextWriterColor.Write(Translate.DoTranslation("There was not enough arguments."));
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                KernelPanic.KernelError(KernelErrorLevel.U, true, 5L, Translate.DoTranslation("Unrecoverable error in argument:") + " " + ex.Message, ex);
            }
        }

    }
}
